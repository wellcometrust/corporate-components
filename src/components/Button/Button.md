# Button

Functional call to action.
