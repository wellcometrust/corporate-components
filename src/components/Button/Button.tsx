import React from 'react';

import Icon from '../Icon/Icon';

type ButtonProps = {
  className?: string;
  children: React.ReactNode;
  disabled?: boolean;
  href?: string | null;
  icon?: string;
  iconPlacement?: '' | 'left' | 'right';
  tabIndex?: string;
  type?: string;
};

export const Button = ({
  children,
  className,
  disabled,
  href,
  icon,
  iconPlacement,
  tabIndex,
  type
}: ButtonProps) => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const Element: any = href ? 'a' : 'button';

  return (
    <Element
      className={className}
      disabled={disabled}
      href={href}
      tabIndex={tabIndex}
      type={!href ? type : null}
    >
      {icon && (iconPlacement === 'left' || iconPlacement === '') && (
        <Icon name={icon} className="btn__icon btn__icon--left" />
      )}
      <span className="btn__text">{children}</span>
      {icon && iconPlacement === 'right' && (
        <Icon name={icon} className="btn__icon btn__icon--right" />
      )}
    </Element>
  );
};

export default Button;
