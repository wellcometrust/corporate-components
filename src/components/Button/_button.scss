%btn-base {
  align-items: center;
  background: transparent;
  border: 0;
  cursor: pointer;
  display: flex;
  font-size: calc(0.94 * var(--text-base-size));
  line-height: 1;
  letter-spacing: rem(0.4);
  padding: 0;
  text-decoration: none;

  &:hover {
    text-decoration: none;
  }
}

.btn {
  @extend %btn-base;
  background: var(--colour-blue-60);
  border-radius: 2px;
  color: var(--colour-white);
  display: inline-flex;
  font-family: var(--font-primary);
  height: calc(5.5 * var(--space-unit));
  justify-content: center;
  min-height: calc(5.5 * var(--space-unit));
  min-width: calc(5.5 * var(--space-unit));
  padding: 0 calc(4 * var(--space-unit));
  transition: background ease 0.4s;

  &:hover {
    background: var(--colour-blue-70);
    color: var(--colour-white);
  }

  &:active {
    background: var(--colour-blue-50);
    color: var(--colour-white);
  }

  &[disabled] {
    background: var(--colour-grey-20);
    color: var(--colour-grey-50);
    cursor: not-allowed;
  }
}

.btn__icon--left {
  margin-left: calc(-2 * var(--space-unit));
  margin-right: var(--space-unit);
}

.btn__icon--right {
  margin-left: var(--space-unit);
  margin-right: calc(-2 * var(--space-unit));
}

.btn--back {
  @extend %btn-nav-base;
  color: var(--colour-blue-80);
  width: 100%;

  @include breakpoint(md) {
    display: none;
  }

  .icon {
    display: flex;
    justify-content: flex-end;
    margin-right: calc(0.5 * var(--space-unit));
    transform: rotate(180deg);
  }
}

.btn--search-mobile {
  @extend %btn-nav-base;
  color: var(--colour-grey-60);

  &:hover {
    color: var(--colour-grey-60);
  }

  .icon {
    margin-left: calc(1.5 * var(--space-unit));
  }

  @include breakpoint(md) {
    color: var(--colour-white);
    margin-right: calc(-4 * var(--space-unit));

    &:hover {
      color: var(--colour-white);
    }
  }

  &.is-disabled {
    opacity: 0;
  }

  &.keyboard-nav {
    color: var(--colour-white);
    height: auto;
    min-height: 0;
    margin-right: calc(2 * var(--space-unit));
    max-width: 0;
    opacity: 0;
    padding-left: 0;
    padding-right: 0;
    pointer-events: none;
    transition: max-width ease 0.4s, opacity ease 0.4s;
    visibility: hidden;

    @include breakpoint(md) {
      display: none;
    }
  }

  &.keyboard-nav:focus {
    max-width: 100px;
    opacity: 1;
    transition: max-width ease 0.4s, opacity ease 0.4s;
    visibility: visible;
  }
}
