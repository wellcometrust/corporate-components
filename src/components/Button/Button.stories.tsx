import React from 'react';
import { storiesOf } from '@storybook/react';
import { boolean, text, select } from '@storybook/addon-knobs';

import Button from './Button';
import Readme from './Button.md';

const ButtonExample = () => {
  const buttonText = text('button text', 'Click me');
  const disabled = boolean('disabled', false);
  const href = text('href', '');
  const icon = select(
    'icon',
    [
      '',
      'arrowNarrow',
      'circleCross',
      'close',
      'download',
      'emailAddress',
      'externalLink',
      'facebook',
      'fax',
      'linkedin',
      'message',
      'phone',
      'search',
      'shareLink',
      'tag',
      'twitter',
      'youTube'
    ],
    ''
  );
  const iconPlacement = select('iconPlacement', ['left', 'right'], '');

  return (
    <Button
      disabled={disabled}
      href={href}
      icon={icon}
      iconPlacement={iconPlacement}
    >
      {buttonText}
    </Button>
  );
};

const stories = storiesOf('Components|Button', module);

stories.add('Button', ButtonExample, { readme: { sidebar: Readme } });
