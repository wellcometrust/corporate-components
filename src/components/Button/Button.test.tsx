import React, { ComponentProps } from 'react';
import { screen, render } from '@testing-library/react';
import user from '@testing-library/user-event';
import Button from './Button';

const setup = (props: Partial<ComponentProps<typeof Button>> = {}) => {
  const defaultProps: ComponentProps<typeof Button> = {
    children: 'Some button',
    ...props
  };

  return render(<Button {...defaultProps} />);
};

describe('<Button />', () => {
  it('renders button', () => {
    const { container } = setup({});
    expect(container).toMatchSnapshot();
  });
  it('allow onClick handler', () => {
    const onClick = jest.fn();
    setup({ onClick });

    const btn = screen.getByRole('button', { name: /Some button/ });

    expect(onClick).not.toHaveBeenCalled();
    user.click(btn);
    expect(onClick).toHaveBeenCalledTimes(1);
  });
  it('can specify href and therefore change base component from button to link', () => {
    setup({ href: 'http://google.com' });

    const btn = screen.getByRole('link', { name: /Some button/ });
    expect(btn).toHaveAttribute('href', 'http://google.com');
  });
});
