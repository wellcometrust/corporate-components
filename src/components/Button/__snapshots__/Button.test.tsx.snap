// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> renders button 1`] = `
<div>
  <button
    class="btn btn--primary"
    type="button"
  >
    <span
      class="btn__text"
    >
      Some button
    </span>
  </button>
</div>
`;
